tidal-link/src/hs/Sound/Tidal/Clock.hs

```diff
@@ -0,0 +1,341 @@
+module Sound.Tidal.Clock where
+
+import qualified Sound.Tidal.Link as Link
+import qualified Sound.Osc.Fd as O
+
+import Control.Concurrent
+import Control.Monad         (when)
+import Control.Monad.Reader  (ReaderT, runReaderT, ask)
+import Control.Monad.State   (StateT, liftIO, evalStateT, get, put, modify)
+
+import Foreign.C.Types (CDouble (..))
+import Data.Int (Int64)
+import Data.Coerce (coerce)
+import System.IO (hPutStrLn, stderr)
+
+type Time = Rational
+
+-- | representation of a tick based clock
+type Clock
+  = ReaderT ClockMemory (StateT ClockState IO)
+
+-- | internal read-only memory of the clock
+data ClockMemory
+  = ClockMemory
+  {clockConfig :: ClockConfig
+  ,clockRef :: ClockRef
+  }
+
+-- | internal mutable state of the clock
+data ClockState
+  = ClockState
+  {ticks  :: Int64
+  ,start  :: Link.Micros
+  ,nowArc :: (Time, Time)
+  ,nudged :: Double
+  } deriving Show
+
+-- | reference to interact with the clock, while it is running
+data ClockRef
+  = ClockRef
+  {rAction :: MVar ClockAction
+  ,rAbletonLink :: Link.AbletonLink
+  }
+
+-- | configuration of the clock
+data ClockConfig
+  = ClockConfig
+  {cQuantum :: CDouble
+  ,cBeatsPerCycle :: CDouble
+  ,cFrameTimespan :: Double
+  ,cEnableLink :: Bool
+  ,cSkipTicks :: Int64
+  ,cProcessAhead :: Double
+  ,cTickAction :: TickAction
+  }
+
+-- | action to be executed on a tick,
+-- | given the current timespan and nudge
+type TickAction
+  = (Time,Time) -> Double -> LinkOperations -> IO ()
+
+-- | link operations for easy interaction with the clock
+data LinkOperations
+  = LinkOperations
+  {timeAtBeat    :: Link.Beat -> IO Link.Micros
+  ,timeToCycles  :: Link.Micros -> IO Time
+  ,getTempo      :: IO Link.BPM
+  ,setTempo      :: Link.BPM -> Link.Micros -> IO ()
+  ,linkToOscTime :: Link.Micros -> O.Time
+  ,beatToCycles  :: CDouble -> CDouble
+  ,cyclesToBeat  :: CDouble -> CDouble
+  }
+
+-- | possible actions for interacting with the clock
+data ClockAction
+  = NoAction
+  | SetCycle Time
+  | SetTempo Time
+  | SetNudge Double
+  deriving Show
+
+defaultCps :: O.Time
+defaultCps = 0.575
+
+defaultConfig :: ClockConfig
+defaultConfig = ClockConfig
+              {cFrameTimespan = 1/20
+              ,cEnableLink = True
+              ,cProcessAhead = 3/10
+              ,cSkipTicks = 10
+              ,cQuantum = 4
+              ,cBeatsPerCycle = 4
```

Comment:
It's mostly redundant. I thought it could possibly be useful for some people but wasn't sure if it would be used. I was thinking people think of cycles in interesting and different ways and that I don't want to impose my way. It could also act as a workaround for the limits of Link BPM.